python/rpdk/java/templates/init/guided_aws/TagHelper.java

```diff
@@ -0,0 +1,250 @@
+     *
+     * Note: Like convertToMap, convertToSet filters out value-less tag entries.
+     *
+     * @param tagMap Map of tags to convert
+     * @return Set of Tag objects
+     */
+    public static Set<Tag> convertToSet(final Map<String, String> tagMap) {
+        if (MapUtils.isEmpty(tagMap)) {
+            return Collections.emptySet();
+        }
+        return tagMap.entrySet().stream()
+            .filter(tag -> tag.getValue() != null)
+            .map(tag -> Tag.builder()
+                .key(tag.getKey())
+                .value(tag.getValue())
+                .build())
+            .collect(Collectors.toSet());
+    }
+
+    /**
+     * shouldUpdateTags
+     *
+     * Generate tags to put into resource creation request.
+     * This includes user defined tags and system tags as well.
+     */
+    public final Map<String, String> generateTagsForCreate(final ResourceModel resourceModel, final ResourceHandlerRequest<ResourceModel> handlerRequest) {
+        final Map<String, String> tagMap = new HashMap<>();
+
+        if (handlerRequest.getSystemTags() != null) {
+            tagMap.putAll(handlerRequest.getSystemTags());
+        }
+
+        if (handlerRequest.getDesiredResourceTags() != null) {
+            tagMap.putAll(handlerRequest.getDesiredResourceTags());
+        }
+
+        // TODO: get tags from resource model based on your tag property name
+        // TODO: tagMap.putAll(convertToMap(resourceModel.getTags()));
+        return Collections.unmodifiableMap(tagMap);
+    }
+
+    /**
+     * shouldUpdateTags
+     *
+     * Determines whether user defined tags or system tags have been changed during update.
+     */
+    public final boolean shouldUpdateTags(final ResourceModel resourceModel, final ResourceHandlerRequest<ResourceModel> handlerRequest) {
+        return areResourceTagsChanging(resourceModel, handlerRequest) || areSystemTagsChanging(handlerRequest);
+    }
+
+    private boolean areResourceTagsChanging(final ResourceModel resourceModel, final ResourceHandlerRequest<ResourceModel> handlerRequest) {
+        final Map<String, String> currentTags = getCurrentTags(handlerRequest);
+        final Map<String, String> desiredTags = getNewDesiredTags(resourceModel, handlerRequest);
+        return ObjectUtils.notEqual(currentTags, desiredTags);
+    }
+
+    private Map<String, String> getCurrentTags(final ResourceHandlerRequest<ResourceModel> handlerRequest) {
+        final Map<String, String> currentTags = handlerRequest.getPreviousResourceTags();
+        // TODO: get tags from previous resource state based on your tag property name
+        // TODO: currentTags.putAll(handlerRequest.getPreviousResourceState().getTags());
+        return currentTags;
+    }
+
+    private Map<String, String> getNewDesiredTags(final ResourceModel resourceModel, final ResourceHandlerRequest<ResourceModel> handlerRequest) {
+        final Map<String, String> desiredTags = handlerRequest.getDesiredResourceTags();
```

Comment:
not including stack-level tags?

Reply:
I thought handlerRequest.getDesiredResourceTags() are stack-level tags.